lgc/builder/ArithBuilder.cpp

```diff
@@ -355,6 +355,25 @@ Value *BuilderImpl::aSinACosCommon(Value *x, Constant *coefP0, Constant *coefP1)
   return CreateFMul(sign, result);
 }
 
+// =====================================================================================================================
+// Common code for find smsb and count leading sign bits
+//
+// @param x : Input value X
+// @param isCountLeadingSignBits : if count leading sign bits
+Value *BuilderImpl::findSMsbOrCountLeadingSignBitsCommon(Value *value, bool isFindSMsb, const Twine &instName) {
+  assert(value->getType()->getScalarType()->isIntegerTy(32));
+
+  Constant *negOne = ConstantInt::get(value->getType(), -1);
+  Value *result =
+      scalarize(value, [this](Value *value) { return CreateUnaryIntrinsic(Intrinsic::amdgcn_sffbh, value); });
+  if (isFindSMsb)
+    result = CreateSub(ConstantInt::get(value->getType(), 31), result);
+  Value *isNegOne = CreateICmpEQ(value, negOne);
+  Value *isZero = CreateICmpEQ(value, Constant::getNullValue(value->getType()));
+  Value *isNegOneOrZero = CreateOr(isNegOne, isZero);
+  return CreateSelect(isNegOneOrZero, negOne, result, instName);
+}
```

Comment:
Looking again at the ISA docs, v_ffbh_i32 / v_cls_i32 (which the intrinsic maps to) is already defined as producing -1 if the input is 0 or -1. So:

* In the `!isFindSMsb` case, none of the extra checks are required, the output of amdgcn_sffbh can be used directly
* In the `isFindSMsb` case, I believe that we likely get better codegen on balance by changing the logic to:
```
  Value *isNegOne = CreateICmpEQ(result, negOne);
  result = CreateSub(ConstantInt::get(value->getType(), 31), result);
  return CreateSelect(isNegOne, negOne, result, instName);
```

Overall, I suggest removing the new helper method and instead using the simpler method directly in CreateCountLeadingSignBits, and then implement CreateFindSMsb by calling CreateCountLeadingSignBits.

Reply:
Thanks for pointing, code updated.